kpatch-build/create-diff-object.c

```diff
@@ -2303,6 +2303,23 @@ static int kpatch_is_core_module_symbol(char *name)
 		!strcmp(name, "kpatch_shadow_get"));
 }
 
+#ifdef __x86_64__
+static int may_need_dynrela(const struct rela *rela)
+{
+	if (!rela->sym->sec)
+		return 1;
+
+	return (rela->sym->type == STT_FUNC &&
+		rela->type == R_X86_64_32S &&
+		!strchr(rela->sym->name, '.'));
+}
```

Comment:
Will this inadvertently affect function symbols like `aio_setup_rw.isra.17` or `percpu_ref_get_many.constprop.21`?  I don't have any examples off hand, but I know I've seen the compiler generate names like that in the past.  (I haven't followed this PR closely, so apologies if such names don't apply here.)

Reply:
I had the same question some time ago, so I checked a few kernels from RHEL, VZ, OpenSUSE and Ubuntu - all functions with dots in their names that were used in the relocations of such kind (R_X86_64_32S) were nested ones. GCC seems currently not to use the optimized copies of the functions (*isra*, *constprop*, ...) when an address of a callback is needed.

Of course, I cannot be 100% sure this is the case for all kernels and all GCC versions. Ingenuity of the compilers should not be underestimated.

Still, this patch covers all problematic cases I have encountered so far. Even if some non-nested function with a dot is used as a callback, the patch does not make things worse than they were before it.

Reply:
Thanks for the explanation and due diligence research.  One more high level question: if kpatch patches a static local callback named "foo" and there are several other "foo" function symbols scattered across vmlinux/modules, would this change resolve "foo" to the appropriate object file?  (ie, the original implementation of the instance that kpatch modified.)

Also, this shouldn't hold up the patch, but is it possible to convert any of the input patch repros into a non-crashing integration test?

Reply:
>  One more high level question: if kpatch patches a static local callback named "foo" and there are several other "foo" function symbols scattered across vmlinux/modules, would this change resolve "foo" to the appropriate object file?

If I understand the dynrela handling code correctly, it should be OK.
There are 2 cases here:

1. There are 2 or more 'foo' functions and all these are changed by the patch. KPatch/Livepatch core will use both function name and its index ('sympos') in the list of such symbols in the patch module to find the right 'foo'.
2. One 'foo' function is changed by the patch and another one is not. The relocations for the former will be resolved the same way as in (1). For the second 'foo' function, KPatch/Livepatch core will assume it is exported by the kernel or a module and will search for it appropriately.

Reply:
> Also, this shouldn't hold up the patch, but is it possible to convert any of the input patch repros into a non-crashing integration test?

It might be possible but, perhaps, quite cumbersome.

One could prepare a patch that adds nops in `create_user_ns()` and `free_user_ns()` (kernel/user_namespace.c), for example. Build the binary patch and check its binary code, e.g. with objdump or libelf-based app. Find the place where the address or `free_user_ns()` is loaded somewhere (this may be tricky) and check if there is a direct relocation of type R_X86_64_32S there. If there is one, the test fails.

If there is no direct relocations there, the test may try to load the patch module, do some unshare / clone(NEW_USER) operations, to make sure the patched code is at least called.

Something like that. I am not familiar with the integration test suite of KPatch yet, so  I cannot say if this can be implemented here.

Reply:
> if we move these lines up into the search for STT_FILE, there would only be a single assignment to hint, like hint = strdup(sym->name);
Good point. Will do it.

Reply:
> how about the attached .patch file? The idea here being that adding the function pointer should be enough to generate the relocation. (Ie, I don't think we even need to call it.)

You are right.

At least on my system, the same kind of relocation is created for `fp = &cmdline_proc_show;` in your test as for the callback assignements in my patches.

I have checked that the different addresses are found and the test fails if the patch is built using the original create-diff-object,  without the fix. The test passes if the fixed create-diff-object is used.

Reply:
👍 seems like a reasonable change, at least for x86 (I'm not very knowledgeable about other arches we'd (like to) support, though.)  We can add integration tests in a follow-up commit.

I'll give @jpoimboe some time to review in case he still wants to comment.

Reply:
Apologies, spectre/meltdown has me pulled in other directions.  I hope to get to this soonish.